### DISCUSSION TITLE
ROUGHNESS from Monin-Obukhov Z_0; numerical instability with &WIND

### DISCUSSION THREAD

**Discussion by obscureed** [firemodels/fds#13375] [Q&A] [answered]
I'm doing a simulation of a wind profile around a building, using the Monin-Obukhov (M-O) profile in `&WIND`. My question is about the `ROUGHNESS` parameter applied to the ground `&SURF`. This is discussed in Section 16.3.4 of the FDS User Guide, including this: "This parameter has a more intuitive meaning than its counterpart in M-O similarity theory, z_0, because it can be viewed as the characteristic height of ground level obstructions." There is an equation that the sand-grain `ROUGHNESS`, with symbol s, is 32.6 times larger than the M-O roughness length, symbol z_0.

Something is slightly strange here. With "rough" terrain, described as "high crops, scattered obstacles such as trees or hedgerows, vineyards", the suggested value of z_0 is 0.25m, so s is 8.2m. The characteristic height of trees and hedgerows is 8m? For "roughly open" terrain, with z_0=0.1m, the characteristic height of "low crops and occasional obstacles (single bushes)" is 3m? These seem like large characteristic heights compared to the objects. Should I accept that these are just parameters in wall functions, and that "characteristic height" does not have to be the same as "typical height"? I note that all of the Validation examples use ROUGHNESS less than 1.0m, often much less, and only a couple of Verification examples go above 1.0m.

I'm asking partly to understand the situation, but also because I sometimes see numerical instability when running these models. (`ERROR: Numerical Instability - FDS stopped`.) Some of this appears to arrive with the initialization, so I can often fix it by (a) starting with zero wind speed and ramping it up in the first 10s, and (b) using `&INIT` to specify temperature and density. Sometimes even these measures do not work, maybe in particular with fine mesh resolution.

An example that diverges is the attached .fds file (1213-prelim-fine.fds.txt below). This was intended to be a "finer than usual" test in a mesh-dependence study. There are several mesh blocks, with several changes in resolution, though never more than a factor of 2 between neighbors. There are some non-cubic cells, though never more than 2:1 aspect ratio. This is an inconveniently large model for troubleshooting, but it does have the convenient feature of diverging in the first two timesteps. This case diverges quickly, but it runs OK with a slightly lower value of `ROUGHNESS`, even 5.0m instead of 8.2m.

I've also attached below a less fine .fds file -- it just doubles the cell size in all directions for all mesh blocks. With the `&RAMP` and the `&INIT` that I mentioned, this runs well. Without the `&RAMP`, it runs, but pressure iterations are constantly maxed out at 10. It is probably heading for a later numerical instability (though I have not checked this specific example) -- even if it doesn't crash, any solution that constantly maxes out pressure iterations is probably junk.

I'd welcome any comments on how to set up these models to run stably. I tried a small initial timestep, but it still diverged once the timestep increased. I tried changing from "neutral" `L=1000000.0` to "stable" `L=350.0`. I tried a couple of different definitions of the thermal BC of the ground.

Two more points:
• Am I able to initialize velocity throughout the fluid? Could it help?
• I tried adding `NOISE=.FALSE.` or `NOISE_VELOCITY=0.001` to the `&INIT` line, but they both refused to run: `ERROR(101): Problem with INIT number 1`. I think these options are in `&MISC`, and the User Guide is wrong to list them in `&INIT`.

[1213-prelim-fine.fds.txt](https://github.com/user-attachments/files/16815629/1213-prelim-fine.fds.txt)
 
[1213-prelim-lessfine.fds.txt](https://github.com/user-attachments/files/16816009/1213-prelim-lessfine.fds.txt)

**Comment by rmcdermo** [comment]
Thanks.  We will take a look at these.

**Comment by rmcdermo** [comment]
@obscureed I do not understand setting both TEMPERATURE and DENSITY on the INIT line.  These should be redundant.  And the latest code only allows TEMPERATURE, which should be handled with TMPA.  Anyway, this is just FYI... I'm starting to look at the case.

**Reply by obscureed** [reply]
The models I have shared were just initial testing. The full models use the same building, but with some equipment releasing air at various temperatures. From the start, some of these models contained small regions of unrealistically low temperatures -- lower than ambient, and lower than any outlet temperatures from the equipment. These regions did not disperse -- in fact, they became worse and plateaued at -230 degC. Unsurprisingly, this was not a healthy solution: pressure iterations maxed out at 10, and eventually the solver gave up with numerical instability. Anyway, to tackle this, I applied the 10-second `&RAMP` to windspeed and the `&INIT` to both temperature and density. I suspect that the biggest effect was the `&RAMP`, but I did both as a belt-and-braces approach and did not test them separately.

**Comment by rmcdermo** [comment]
OK, I'm seeing the same as you: instability after 2 steps with latest source.

Need to think about this.  A sand grain roughness larger than the first DZ has always seemed pointless to me.  To be honest, I have a lot of questions about applying MO theory in FDS.   We resolve all the physics that requires you to need MO to begin with.

**Reply by obscureed** [reply]
Maybe we should not think about the sand grain roughness parameter as representing enormous sand grains -- it is just a parameter in a wall function.

I have seen this before, trying to simulate atmospheric boundary layers (ABLs) in general-purpose CFD software (k-epsilon or k-omega RANS as the turbulence model). I applied a profile to represent the ABL at the upwind boundary. If I applied standard wall functions to the ground BC, the ABL profile very quickly adjusted itself to something nearly uniform. (One result of this was that the results were sensitive to the height of the model. The more height, the greater the average inlet speed. This average inlet speed is then present at building heights a short distance after the boundary. The windspeed that the building experiences is only distantly related to the windspeed specified at the boundary.) To avoid this, it helped to apply some extreme wall functions to the ground. See Hargreaves and Wright, "The use of commercial CFD software to model the atmospheric boundary layer". I never did work through what the extreme wall functions were doing, but I think maybe they added in much more k and much more epsilon-or-omega into the RANS model, as well as adjusting the wall shear stress.

In theory, it might be possible to maintain the ABL profile by modelling the real terrain. Experiments in wind tunnels sometimes try to do this with patterns of cubes screwed to the floor. Building cubes into an FDS model would not be onerous, but it would need validation -- I do not have confidence that it would inevitably resolve itself into the correct physics at any sensible mesh resolution.

So we are back to applying weird boundary conditions for the ground. The recommended approach in FDS, with strange-looking `ROUGHNESS` values, seems to maintain the ABL profile fairly well. (The RANS models I mentioned were years ago, so I have not compared them.) But not when it fails to solve...

**Comment by rmcdermo** [comment]
I started digging into the literature on this.  [This paper](https://link.springer.com/article/10.1007/s10546-016-0225-y) is the tip of the iceberg.  But essentially, it is not recommended to use MO if you first grid cell is smaller than the sand grain roughness.  What I would like to do is find a good way to gracefully transition to something reasonable.  Anecdotally, over the years I've noticed that simply using NO_SLIP for outdoor flows tends to recover experimentally measured profiles better.

For testing, I can of course use the verification cases we already have, but they may not be extreme enough.  What should we be comparing agains in your case here?  The MO profile will not be maintained around a structure, so that is not a good comparison.  I'm open to suggestions.

**Reply by obscureed** [reply]
Thanks for finding that paper and that iceberg. The paper says: "Incorporation of such a formulation in contemporary LES codes should not be a challenging task," but it's still far beyond my skills/budget/timescale. The focus of my current models is the flow over the top of the building, and if that's reasonable, then what's happening at the ground may not be critical. I'll rerun some cases with NO_SLIP on the ground -- thanks for this suggestion. If the conclusions don't change markedly, that's enough testing for this project. If that works, I'll mark your suggestion as an answer (pending further research) and make my exit. Thanks for your help!

**Reply by rmcdermo** [reply]
Glad you were able to access it.  I would not expect you to be responsible for implementing the roughness corrections.  I'm working on it to see if I can get something that is simple and does not break what we have working to this point.  I'm interested to see how NO_SLIP works for you.

**Reply by obscureed** [reply]
OK, I've re-run my mesh-dependence test cases with NO_SLIP=T instead of ROUGHNESS=8.15. (I added an &INIT line to the new cases as well.) All these cases ran successfully, unlike the previous cases (which I'll label "M-O", Monin-Obukhov), where the finest z-resolutions had numerical instability. Actually, they are not all mesh-dependence tests; some of them change other parameters such as the extent of the fluid domain.

As I've mentioned, the focus of these models is the "well" at the top of the building: a flat roof surrounded by the top storey plus a parapet. Therefore some quantities of interest in the test models (with nothing in the well) are:
• Volume-average speed in the well.
• Exchange flowrate between the well and the exterior.
• Volume flowrate passing over the top of the building. (In the .fds file, I set up volume flow &DEVCs for the entire width of the building, extending to 4m and 8m above the building roof. Actually, I think that only 3m and 7m of these surfaces have flow, because of the parapet -- it does not really matter, as it is all comparative. The wind that is blocked by the building has a choice whether to go around or go over. The split of how much goes around and how much goes over might not be very mesh-dependent, but it might be affected by the overall width and height of the modelled fluid.)

These quantities (averaged over a fairly long time, from t=600s to t=1800s) are plotted for a variety of cell counts in the graphs below.
• The setups of the "no-slip" and "M-O" cases with equal numbers of cells are identical apart from the ground BC and the `&INIT`.
• I am happy with this as a mesh-dependence study (for the level of accuracy that is justified for this project). The new results with high cell counts spoil the plateaus slightly, and suggest that there could still be long-term trends, but all the results except for the coarsest cases fit in a +/-10% band.
• Also, the differences between M-O and no-slip results are small. This applies only to the specific postprocessing that I'm looking at, which is not expected to be very sensitive to the ground BCs. The no-slip BC could be preferred, for example because of stability.
• I think that this insensitivity to the ground BCs may mean that this is not a good test case for your changes to the code (though the stability is encouraging). I'm afraid I do not have any more time to devote to this for now.

I'll leave this open for a day or two, in case any other discussion points crop up, then mark it as answered.
Thanks for your help!


![image](https://github.com/user-attachments/assets/872e365e-3528-44a3-95d9-33843f620fc2)

![image](https://github.com/user-attachments/assets/c72ef904-a6e7-4119-ad2a-002a9cb42c10)

![image](https://github.com/user-attachments/assets/55507263-0c62-44fa-b889-43914e074319)

**Comment by rmcdermo** [comment, marked as answer]
Based on this thread, I've made a change to the code that now automatically converts to using `NO_SLIP_BC` if the local wall cell normal distance is less than the specified sand grain roughness for that surface.  This is equivalent to the user setting `NO_SLIP=T` for that surface.  I've added discussion and a warning to the user guide.  See PR #13376 

The way to think about `Z_0` is that it is the "y intercept" of the profile in a semilogy plot.  Below I have 4 plots considering 2 different grid spacings 0.25 m (your case) and 25 m.  On the left you see the physical space profile for the MO parameters of your case.  On the right is the semilogy plot.  Essentially, you can see that a linear profile (i.e., `NO_SLIP`) is a pretty reasonable approximation to the well-resolved case.

![Screenshot 2024-08-30 at 5 17 40 PM](https://github.com/user-attachments/assets/04b7f184-fd28-4b8c-bc40-e9d5c8ea4014)

I'm running your case now and it seems to be doing ok.  Note that unless there is a very good reason for it, I would get rid of setting `CONVECTIVE_HEAT_FLUX=0`.  Let me know how it goes.

**Reply by obscureed** [reply]
I copied `CONVECTIVE_HEAT_FLUX=...` from the LNG cases in the Validation suite. Those cases applied small-looking values, so I just applied zero as a convenient assumption. I could not work out what thermal conditions were in force for other Validation cases. I did not want to have to decide and justify thermal properties of soil, etc. I now have several cases that I do not want to rerun unless necessary, so I'll stick with that BC in the current project (unless you think it's actually wrong enough to change the conclusions). For future projects, what would you suggest instead?

I now see that some other Validation cases (Ekman) use `NET_HEAT_FLUX=...` and then small-looking values, depending on thermal stability profiles. For neutral stability, would `NET_HEAT_FLUX=0.0` be better than `CONVECTIVE_HEAT_FLUX=0.0`? In the absence of radiation, I would expect them to be similar.

**Reply by rmcdermo** [reply]
In absence of radiation, yes, they should be identical.  But I really do not like trying to apply this BC locally in an LES.  It is only meant to satisfy the mean dT/dz for the MO profile.  Locally, dT/dz is always varying and trying to enforce it locally in space and time leads to unphysical results and often instabilities.  It is much better to try to apply the mean temperature to the surface and let the calculation proceed.  If you have a known external flux, then I will apply this with EXTERNAL_FLUX.

**Comment by rmcdermo** [comment]
Glad to hear that your cases are running.  The one I launched is also still going, so I think this is a reasonable approach.

One thing I noticed is that the downstream boundary cuts through the wake of the building (see image below).  I will let you decide how much this matters for the dip in the roof.  I suspect you have already seen the paper by Wojciech Wegrzynski, [Wind and Fire Coupled Modelling -- Part II: Good Practice Guidelines](https://link.springer.com/article/10.1007/s10694-018-0749-4).  I think it is a very nice paper.  In fig 1, you will see that for computational wind engineering he suggests A_block/A_cross to be about 3% and the downstream boundary to be more than 10H behind the building.  So, I would not draw too many conclusions about MO vs. NO_SLIP until we satisfy these criteria.  But maybe for your project this is overkill.

![Screenshot 2024-09-03 at 11 08 28 AM](https://github.com/user-attachments/assets/81b2810a-da91-4ec5-80e3-e40ffae56bf2)

**Reply by obscureed** [reply]
I think I have satisfied those guidelines. The wake downwind is from y=26 to y=168, so distance 142m, which is approximately 11H. The building represents a blockage H13m × W42m, and the fluid domain is H92m × W208m, so the blockage is just under 3%. (These guidelines are not that hard to satisfy for a squat building like the current one. For a skyscraper, they are much more painful.) I had not seen the paper that you mentioned, but there are various other guidelines (not involving fires) that are similar:
• the COST Action 732 guidelines (https://sciencespo.hal.science/ENPC-CEREA/hal-04181390v1);
• some Japanese guidelines (doi:10.1016/j.jweia.2008.02.058);
• some papers by Prof Blocken (https://doi.org/10.1016/j.buildenv.2015.02.015).

I haven't seen any surges of wake flow as big as the one in the image you've shown. I have some `DEVC` monitors that tell me that the flow out of the downwind boundary is reassuringly all outwards; zero backflow. Perhaps I should have left them in the .fds file I shared, but I thought they might clutter the picture. Things like this:
```
&VENT XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, SURF_ID='OPEN' /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, ID='volflow_vent_wind__6',  QUANTITY='V-VELOCITY', SPATIAL_STATISTIC='AREA INTEGRAL',  /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, ID='volflowplus_vent_wind__6',  QUANTITY='V-VELOCITY', SPATIAL_STATISTIC='AREA INTEGRAL', QUANTITY_RANGE(1)=0.0, /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, ID='volflowminusvent_wind__6',  QUANTITY='V-VELOCITY', SPATIAL_STATISTIC='AREA INTEGRAL', QUANTITY_RANGE(2)=0.0, /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, QUANTITY='TEMPERATURE', ID='tave_vent_wind__6',  STATISTICS='MEAN',  /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, QUANTITY='TEMPERATURE', ID='tmax_vent_wind__6',  STATISTICS='MAX',  /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, STATISTICS='AREA INTEGRAL',  QUANTITY='V-VELOCITY', ID='areaint_posvel_vent_wind__6',  QUANTITY_RANGE= 0.0,999.9,  /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, STATISTICS='AREA INTEGRAL',  QUANTITY='V-VELOCITY', ID='areaint_negvel_vent_wind__6',  QUANTITY_RANGE= -999.9,0.0,  /
&DEVC XB= -96.000, 112.000, 168.000, 168.000,   0.000,  92.000, STATISTICS='AREA INTEGRAL',  QUANTITY='V-VELOCITY', ID='areaint_netvel_vent_wind__6',  /
```
The equivalent `DEVC` monitors on the top and side show that the flow there is nicely behaved: some inflow, but (after t=200s) very little compared to the overall flows and sizes.

I do change mesh resolution fairly often -- this takes a while to set up, but I think it's the only way to keep the cell count manageable. There are definitely some mesh/mesh flow artefacts in the solution -- not literally checkerboarding (in the sense of fluctuations between pairs of neighboring cells), but artefacts at mesh/mesh boundaries. For example, see the yellow/red pattern in the image below. These artefacts are less noticeable where there are actual flow features happening (in other words, anywhere interesting), so I ignore them.
![checker_grid](https://github.com/user-attachments/assets/ec04e023-8093-45e2-96a9-7415aaa653d6)

**Reply by rmcdermo** [reply]
I don't have exact numbers, but visually it looks like your downstream boundary is about 6H behind the building (H being the height of the building).  I think the paper suggests this distance to be 10-15H.  Anyway, I get that these calcs are expensive and you have to cut somewhere.  And it sounds like you are well aware of the guidance, so I won't belabor.  I think things are headed the right direction.

**Reply by obscureed** [reply]
In the image I just posted, flow is from left to right (it's the positive y-direction), and the pale grey rectangle near the left is the building. (The black above it is fine mesh.) Honestly, the length of the fluid from the downwind side of the building to the down boundary is 10.9*H, where H=13m includes the parapet. If you're sure you're seeing 6H, we need to check we're looking at the same model.

**Reply by rmcdermo** [reply]
Sorry, yeah are right, I was looking at the wrong height.  Anyway, given that the wake is clearly interacting with the boundary, I would use caution here.  Vectors are more helpful than the contours at seeing this.
